### PR TITLE
Rename --provision-config parameter to --channels

### PR DESCRIPTION
**Pull request by TomasHofman**
...and allow combination with --repositories

Resolves https://github.com/wildfly-extras/prospero/issues/152

Work in progress, do not merge yet.

### CODE REVIEW

File: examples/wildfly-core-channel.yaml [hunk 1]
```diff
@@ -0,0 +1,10 @@
+---
+schemaVersion: "2.0.0"
+repositories:
+  - id: "central"
+    url: "https://repo1.maven.org/maven2/"
+  - id: "jboss-public"
+    url: "https://repository.jboss.org/nexus/content/groups/public/"
+manifest:
+  url: "file:///home/thofman/Projects/prospero/examples/wildfly-core-manifest.yaml"
```
Comment (spyrkob) [inline examples/wildfly-core-channel.yaml:9]: I guess the github url will be in the final PR?

File: prospero-common/src/main/java/org/wildfly/prospero/actions/ProvisioningAction.java [hunk 2]
```diff
@@ -57,26 +60,23 @@ public ProvisioningAction(Path installDir, MavenSessionManager mavenSessionManag
...
-     * @param config
-     * @param channels
-     * @throws ProvisioningException
-     * @throws MetadataException
+     * @param provisioningDefinition prospero provisioning definition
      */
-    public void provision(ProvisioningConfig config, List<Channel> channels) throws ProvisioningException, OperationException {
-        doProvision((pm,options)->pm.provision(config, options), enforceChannelNames(channels));
-    }
+    public void provision(ProvisioningDefinition provisioningDefinition)
```
Comment (spyrkob) [inline prospero-common/src/main/java/org/wildfly/prospero/actions/ProvisioningAction.java:67]: I think I'd rather keep this using ProvisioningConfig and channels directly rather then ProvisioningDefinition

File: prospero-common/src/main/java/org/wildfly/prospero/api/ArtifactUtils.java [hunk 3]
```diff
@@ -31,25 +34,39 @@ public static int compareVersion(Artifact first, Artifact other) {
...
+                    } else if (gav.length == 3) {
+                        return clazz.getDeclaredConstructor(String.class, String.class, String.class)
+                                .newInstance(gav[0], gav[1], gav[2]);
+                    }
                 }
-            } else {
                 // assume the string is a path
                 try {
-                    return new ChannelManifestCoordinate(Paths.get(urlGavOrPath).toAbsolutePath().toUri().toURL());
+                    return clazz.getDeclaredConstructor(URL.class)
```
Comment (spyrkob) [inline prospero-common/src/main/java/org/wildfly/prospero/api/ArtifactUtils.java:62]: Maybe add a check that the file exists?

File: prospero-common/src/main/java/org/wildfly/prospero/api/ArtifactUtils.java [hunk 4]
```diff
@@ -31,25 +34,39 @@ public static int compareVersion(Artifact first, Artifact other) {
         return new ComparableVersion(first.getVersion()).compareTo(new ComparableVersion(other.getVersion()));
     }
 
-    public static ChannelManifestCoordinate manifestFromString(String urlGavOrPath) {
+    public static ChannelManifestCoordinate manifestCoordFromString(String urlGavOrPath) {
+        return coordinateFromString(urlGavOrPath, ChannelManifestCoordinate.class);
+    }
+
+    public static ChannelCoordinate channelCoordFromString(String urlGavOrPath) {
```
Comment (spyrkob) [inline prospero-common/src/main/java/org/wildfly/prospero/api/ArtifactUtils.java:41]: can you add tests for both `channelCoordFromString` and `manifestCoordFromString` - want to make sure we catch if the constructors change

File: prospero-common/src/main/java/org/wildfly/prospero/api/ProvisioningDefinition.java [hunk 5]
```diff
@@ -51,47 +54,73 @@ public class ProvisioningDefinition {
 
     public static final RepositoryPolicy DEFAULT_REPOSITORY_POLICY = new RepositoryPolicy(true, RepositoryPolicy.UPDATE_POLICY_ALWAYS, RepositoryPolicy.CHECKSUM_POLICY_FAIL);
 
+    /**
+     * Galleon feature pack location. Can be either a well-known name (like "eap-8.0") that references a predefined combination
```
Comment (spyrkob) [inline prospero-common/src/main/java/org/wildfly/prospero/api/ProvisioningDefinition.java:58]: eap-8.0/wildfly

File: prospero-common/src/main/java/org/wildfly/prospero/api/ProvisioningDefinition.java [hunk 6]
```diff
@@ -51,47 +54,73 @@ public class ProvisioningDefinition {
...
+     * of Galleon configuration and channels, or a standard Galleon feature pack location string (see
+     * https://docs.wildfly.org/galleon/#_feature_pack_location).
+     */
     private final String fpl;
+
+    /**
+     * This field would only contain channels parsed from the prosper-known-combinations.yaml file ("well-known feature packs"),
```
Comment (spyrkob) [inline prospero-common/src/main/java/org/wildfly/prospero/api/ProvisioningDefinition.java:65]: The comments talk about command line attributes, while the `ProvisioningDefinition` is in prospero-common. I think the ProvisioningDefinition should be moved to cli module

File: prospero-common/src/main/java/org/wildfly/prospero/api/ProvisioningDefinition.java [hunk 7]
```diff
@@ -104,23 +133,29 @@ private boolean channelsMissingManifest() {
...
+                new RemoteRepository.Builder("MRRC", null, "https://maven.repository.redhat.com/ga/").build(),
+                new RemoteRepository.Builder("central", null, "https://repo1.maven.org/maven2").build()
+        );
+    }
+
+    private static List<Repository> extractRepositoriesFromChannels(List<Channel> channels) {
+        return channels.stream().flatMap(c -> c.getRepositories().stream()).collect(Collectors.toList());
+    }
+
+    private void setKnownFeaturePackChannels(List<Repository> overrideRepositories,
```
Comment (spyrkob) [inline prospero-common/src/main/java/org/wildfly/prospero/api/ProvisioningDefinition.java:148]: what happens if channelCoordinates is not empty?

Do you think this method could return List<Channel> and assign this.channels in the constructor?

File: prospero-common/src/main/java/org/wildfly/prospero/api/ProvisioningDefinition.java [hunk 8]
```diff
@@ -104,23 +133,29 @@ private boolean channelsMissingManifest() {
         return channels.stream().anyMatch(c->c.getManifestRef() == null);
     }
 
-    private void setUpBuildEnv(List<Repository> additionalRepositories, Optional<Path> provisionConfigFile,
-                               Optional<ChannelManifestCoordinate> manifestRef, List<Channel> channels) throws IOException {
-        if (!provisionConfigFile.isPresent() && !manifestRef.isPresent()) {
-            if (!additionalRepositories.isEmpty()) {
-                this.channels = TemporaryRepositoriesHandler.addRepositories(channels, additionalRepositories);
+    // TODO: externalize
+    private static List<RemoteRepository> defaultChannelResolutionRepositories() {
```
Comment (spyrkob) [inline prospero-common/src/main/java/org/wildfly/prospero/api/ProvisioningDefinition.java:137]: How about overriding (or at least supplementing) those with `overrideRepositories`?

File: prospero-common/src/main/java/org/wildfly/prospero/api/ProvisioningDefinition.java [hunk 9]
```diff
@@ -164,29 +203,48 @@ public ProvisioningConfig toProvisioningConfig() throws ProvisioningException {
...
+     *
+     * @param versionResolverFactory a VersionResolverFactory instance to perform the channel resolution
+     * @return Channel instances
+     */
+    public List<Channel> resolveChannels(VersionResolverFactory versionResolverFactory) {
+        try {
+            List<Channel> channels = new ArrayList<>(this.channels);
+
+            if (!channelCoordinates.isEmpty()) {
+                channels.addAll(versionResolverFactory.resolveChannels(channelCoordinates, defaultChannelResolutionRepositories()));
```
Comment (spyrkob) [inline prospero-common/src/main/java/org/wildfly/prospero/api/ProvisioningDefinition.java:217]: shouldn't channelCoordinates allow path as well as URL and GAV?

File: examples/wildfly-core-channel.yaml (hunk 1 above)
Comment (TomasHofman) [reply examples/wildfly-core-channel.yaml:9]: Yes, I think the URL to GitHub are best for this use cases because they are absolute and will stay active. Ideally it should also be pointing to the "main" branch.

File: prospero-common/src/main/java/org/wildfly/prospero/api/ProvisioningDefinition.java (hunk 6 above)
Comment (TomasHofman) [reply prospero-common/src/main/java/org/wildfly/prospero/api/ProvisioningDefinition.java:65]: I will just change the attribute references to builder methods. I think this class is useful as part of programmatic API in common.

File: prospero-common/src/main/java/org/wildfly/prospero/api/ArtifactUtils.java (hunk 4 above)
Comment (TomasHofman) [reply prospero-common/src/main/java/org/wildfly/prospero/api/ArtifactUtils.java:41]: Sure, test added.

File: prospero-common/src/main/java/org/wildfly/prospero/api/ArtifactUtils.java (hunk 3 above)
Comment (TomasHofman) [reply prospero-common/src/main/java/org/wildfly/prospero/api/ArtifactUtils.java:62]: I wouldn't do that here. We don't check URLs and GAVs for existence either.

File: prospero-common/src/main/java/org/wildfly/prospero/actions/ProvisioningAction.java (hunk 2 above)
Comment (TomasHofman) [reply prospero-common/src/main/java/org/wildfly/prospero/actions/ProvisioningAction.java]: OK, I wanted to do channel resolution inside the `ProvisioningAction` class, but it turns out it's OK to do it in the `InstallCommand` as well.

File: examples/wildfly-core-channel.yaml (hunk 1 above)
Comment (TomasHofman) [reply examples/wildfly-core-channel.yaml]: Updated to GH urls.

File: prospero-common/src/main/java/org/wildfly/prospero/api/ProvisioningDefinition.java (hunk 8 above)
Comment (TomasHofman) [reply prospero-common/src/main/java/org/wildfly/prospero/api/ProvisioningDefinition.java:140]: I was planning originally to have separate CLI option to control this, but attaching this to `--repositories` as well seems OK.

File: prospero-common/src/main/java/org/wildfly/prospero/api/ProvisioningDefinition.java (hunk 9 above)
Comment (TomasHofman) [reply prospero-common/src/main/java/org/wildfly/prospero/api/ProvisioningDefinition.java:217]: If they were Paths, they were already converted to URLs at this point. The conversion is done in the `ProvisioningDefinition.Builder#setChannelCoordinates()`.

Comment (spyrkob) [reply prospero-common/src/main/java/org/wildfly/prospero/api/ProvisioningDefinition.java]: ah right, missed that 👍

File: prospero-cli/src/test/java/org/wildfly/prospero/cli/commands/InstallCommandTest.java [hunk 10]
```diff
@@ -107,20 +111,20 @@ public void errorIfChannelsIsNotPresentAndUsingCustomFplOnInstall() {
                 CliConstants.FPL, "foo:bar");
         assertEquals(ReturnCodes.INVALID_ARGUMENTS, exitCode);
         assertTrue("output: " + getErrorOutput(), getErrorOutput().contains(String.format(
-                CliMessages.MESSAGES.prosperoConfigMandatoryWhenCustomFpl().getMessage(), CliConstants.PROVISION_CONFIG)));
+                CliMessages.MESSAGES.channelsMandatoryWhenCustomFpl().getMessage(), CliConstants.CHANNELS)));
```
Comment (spyrkob) [inline prospero-cli/src/test/java/org/wildfly/prospero/cli/commands/InstallCommandTest.java:114]: String.format is not needed, the new message doesn't take a parameter

File: prospero-cli/src/test/java/org/wildfly/prospero/cli/commands/InstallCommandTest.java [hunk 11]
```diff
@@ -220,30 +224,33 @@ public void fplAndDefinitionAreNotAllowedTogether() throws Exception {
...
-                .exclusiveOptions(CliConstants.PROVISION_CONFIG, CliConstants.CHANNEL_MANIFEST).getMessage()));
+                .exclusiveOptions(CliConstants.CHANNELS, CliConstants.CHANNEL_MANIFEST).getMessage()));
     }
 
     @Test
-    public void provisionConfigAndRemoteRepoSet() throws IOException {
-        final File provisionConfigFile = temporaryFolder.newFile();
+    public void provisionConfigAndRemoteRepoSet() throws Exception {
+        Path channelsFile = temporaryFolder.newFile().toPath();
+        MetadataTestUtils.prepareChannel(channelsFile, List.of(getClass().getClassLoader().getResource("manifest.yaml")));
```
Comment (spyrkob) [inline prospero-cli/src/test/java/org/wildfly/prospero/cli/commands/InstallCommandTest.java:242]: maybe instead of adding the test manifest, just mock the manifest file here? just empty temporary file should do I think

File: prospero-common/src/main/java/org/wildfly/prospero/actions/ProvisioningAction.java [hunk 12]
```diff
@@ -57,26 +57,22 @@ public ProvisioningAction(Path installDir, MavenSessionManager mavenSessionManag
...
     /**
-     * Installs feature pack defined by {@code fpl} in {@code installDir}. If {@code fpl} doesn't include version,
-     * the newest available version will be used.
+     * Provision installation according to given ProvisioningDefinition.
      *
-     * @param config
-     * @param channels
-     * @throws ProvisioningException
-     * @throws MetadataException
+     * @param provisioningDefinition prospero provisioning definition
```
Comment (spyrkob) [inline prospero-common/src/main/java/org/wildfly/prospero/actions/ProvisioningAction.java:62]: wrong `@param`

File: prospero-common/src/main/java/org/wildfly/prospero/api/ProvisioningDefinition.java [hunk 13]
```diff
@@ -137,6 +143,10 @@ public ProsperoConfig getProsperoConfig() {
         return new ProsperoConfig(channels);
     }
 
+    public List<Repository> getOverrideRepositories() {
```
Comment (spyrkob) [inline prospero-common/src/main/java/org/wildfly/prospero/api/ProvisioningDefinition.java:146]: can we remove this? I don't think it's used anywhere

File: prospero-common/src/main/java/org/wildfly/prospero/api/RepositoryUtils.java [hunk 14]
```diff
@@ -18,18 +18,21 @@
...
-
 public class RepositoryUtils {
     public static Repository toChannelRepository(RemoteRepository r) {
         return new Repository(r.getId(), r.getUrl());
     }
 
     public static RemoteRepository toRemoteRepository(String id, String url) {
         return new RemoteRepository.Builder(id, "default", url)
-                .setPolicy(DEFAULT_REPOSITORY_POLICY)
+                .setPolicy(new RepositoryPolicy(true, RepositoryPolicy.UPDATE_POLICY_ALWAYS, RepositoryPolicy.CHECKSUM_POLICY_IGNORE))
```
Comment (spyrkob) [inline prospero-common/src/main/java/org/wildfly/prospero/api/RepositoryUtils.java:31]: Why the `CHECKSUM_POLICY_IGNORE` policy? don't think we should ignore checksum errors

**Review by spyrkob — CHANGES_REQUESTED** [review]

File: integration-tests/src/test/java/org/wildfly/prospero/it/featurepacks/WildflyFpTest.java [hunk 15]
```diff
@@ -66,10 +66,12 @@ public void testInstallProsperoWithWildfly() throws Exception {
         final Path installedProspero = targetDir.resolve("bin").resolve(ExecutionUtils.isWindows()?"prospero.bat":"prospero.sh");
         assertTrue(Files.exists(installedProspero));
 
-        ExecutionUtils.prosperoExecution(CliConstants.Commands.UPDATE, CliConstants.DRY_RUN,
+        // TODO: This doesn't work due to incompatible wildfly-channel library version. Re-enable after we get back to an
+        //  official wildfly-channel release.
+        /*ExecutionUtils.prosperoExecution(CliConstants.Commands.UPDATE, CliConstants.DRY_RUN,
```
[the remark this answers is not in the document]
Comment (TomasHofman) [reply integration-tests/src/test/java/org/wildfly/prospero/it/featurepacks/WildflyFpTest.java:71]: It was something like "Class ChannelManifest not found". I think I used an old version of `prospero-manifest.yaml` - it only updates when running the build with `-DallTests`, which I forgot to be doing. It works now.

File: integration-tests/src/test/resources/manifests/prospero-manifest.yaml [hunk 16]
```diff
@@ -0,0 +1,166 @@
+---
```
[the remark this answers is not in the document]
Comment (TomasHofman) [reply integration-tests/src/test/resources/manifests/prospero-manifest.yaml:1]: Done.

File: prospero-common/src/main/java/org/wildfly/prospero/api/ProvisioningDefinition.java [hunk 17]
```diff
@@ -178,6 +160,7 @@ public ProvisioningConfig toProvisioningConfig() throws MetadataException, Provi
             try {
                 return GalleonUtils.loadProvisioningConfig(definition);
             } catch (XMLStreamException e) {
+                e.printStackTrace();
```
[the remark this answers is not in the document]
Comment (TomasHofman) [reply prospero-common/src/main/java/org/wildfly/prospero/api/ProvisioningDefinition.java:163]: Yes, forgot to remove after some debugging.

File: examples/README.md [hunk 18]
```diff
@@ -0,0 +1,87 @@
...
+  --manifest examples/wildfly-27.0.0.Alpha2-manifest.yaml \
+  --dir installation-dir
+```
+
+(Note: the `--manifest ...` option would be optional for predefined "eap-*" options.)
+
+## Installation Referencing a Channel
+
+```shell
+./prospero install --fpl org.wildfly.core:wildfly-core-galleon-pack:20.0.0.Beta2 \
```
[the remark this answers is not in the document]
Comment (TomasHofman) [reply examples/README.md:65]: If I remove the version, I think the example is going to stop working in the future (when the core-galleon-pack does changes that would not be reflected by the wf-core manifest versioned in this directory).

Maybe I could stress that this readme is intended to developers, not users? Maybe also point out that the FP version would normally be omitted?

BTW my goal was to have some inputs for immediate testing / demonstration, as there are no available manifests around yet and I always have to go and generate a fresh manifest from Wildfly. (I wanted to use wf-core because it's smaller / faster.)

[the remark this answers is not in the document]
Comment (TomasHofman) [reply examples/README.md:65]: I should fix the 20.0.0.Beta3-SNAPSHOT manifest to 20.0.0.Beta2.

File: examples/README.md [hunk 19]
```diff
@@ -0,0 +1,87 @@
...
+
+Description:
+ * The `--fpl` option defines a feature pack to be installed.
+ * The `--channels` option defines a Wildfly Channel (a *Wildfly Channel* references a *Wildfly Channel Manifest* and
+Maven repositories). 
+
+## Installation Referencing a Channel Manifest and Maven Repositories
+
+```shell
+./prospero install --fpl org.wildfly.core:wildfly-core-galleon-pack:20.0.0.Beta2 \
```
[the remark this answers is not in the document]
Comment (TomasHofman) [reply examples/README.md:78]: OK, I wanted to make this combination valid and then didn't.

Are you OK with making the `--manifest` option valid for all input options (pre-defined fpl, custom fpl, galleon config)? Originally it was valid for pre-defined FPL. I think it would be more transparent behavior.

File: prospero-cli/src/test/java/org/wildfly/prospero/cli/commands/InstallCommandTest.java (hunk 10 above)
Comment (TomasHofman) [reply prospero-cli/src/test/java/org/wildfly/prospero/cli/commands/InstallCommandTest.java:114]: You are right!

File: prospero-cli/src/test/java/org/wildfly/prospero/cli/commands/InstallCommandTest.java (hunk 11 above)
Comment (TomasHofman) [reply prospero-cli/src/test/java/org/wildfly/prospero/cli/commands/InstallCommandTest.java:242]: I can give a non-existing URL as a manifest reference, but the channel file need to exists.

File: prospero-common/src/main/java/org/wildfly/prospero/api/ProvisioningDefinition.java (hunk 13 above)
Comment (TomasHofman) [reply prospero-common/src/main/java/org/wildfly/prospero/api/ProvisioningDefinition.java:146]: Yes.

File: prospero-common/src/main/java/org/wildfly/prospero/api/RepositoryUtils.java (hunk 14 above)
Comment (TomasHofman) [reply prospero-common/src/main/java/org/wildfly/prospero/api/RepositoryUtils.java:31]: Thanks for spotting, another forgotten bit. I was testing resolving manifests from local repo, that didn't have checksums...

File: examples/README.md (hunk 18 above)
[the remark this answers is not in the document]
Comment (TomasHofman) [reply examples/README.md]: How does missing `org.jboss.invocation:jboss-invocation-jakarta` manifest? I don't see that being a wildfly-core dependency.

File: examples/README.md (hunk 19 above)
[the remark this answers is not in the document]
Comment (spyrkob) [reply examples/README.md]: Yes that sounds good @TomasHofman 👍

File: examples/README.md (hunk 18 above)
[the remark this answers is not in the document]
Comment (spyrkob) [reply examples/README.md]: > If I remove the version, I think the example is going to stop working in the future (when the core-galleon-pack does changes that would not be reflected by the wf-core manifest versioned in this directory).

Not sure I follow. Without the version it will install the FP version defined by the manifest - so that will always be the 20.0.0.Beta2 and shouldn't change. Or am I missing something?

> BTW my goal was to have some inputs for immediate testing / demonstration, as there are no available manifests around yet and I always have to go and generate a fresh manifest from Wildfly. (I wanted to use wf-core because it's smaller / faster.)
👍 

> How does missing org.jboss.invocation:jboss-invocation-jakarta manifest? I don't see that being a wildfly-core dependency.

When I changed the core versions to 20.0.0.Beta2 provisioning failed with missing artifact. Maybe it was removed in Beta3

**Review by spyrkob — APPROVED** [review]

**Comment by TomasHofman** [comment]
Seems like CI is finally passing :+1:

**Comment by spyrkob** [comment]
Thanks @TomasHofman !
